File: adapters/python/okf_adapter.py
#!/usr/bin/env python3
"""OKF language adapter stub — Python reference (C08 L75).

Implements the language-agnostic contract in adapters/README.md:
  load(path) -> dict
  validate(doc) -> None (raises ValueError on failure)
  emit(doc) -> str

CLI:
  python okf_adapter.py validate <path.okf.json>
  python okf_adapter.py emit <path.okf.json>

Hermetic: stdlib only. Not a Harbor / agent-eval harness.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any, Mapping, MutableMapping, Sequence

OKF_DIALECT = "1.0"
REQUIRED_SHARED_ENTITY_TYPES = frozenset(
    {
        "intent",
        "acceptance",
        "constraint",
        "resource",
        "state",
        "gate",
    }
)
ALLOWED_RELATION_TYPES = frozenset(
    {
        "verified_by",
        "bounded_by",
        "grounds",
        "requires",
        "asserts",
    }
)


def load(path: str | Path) -> MutableMapping[str, Any]:
    """Read an OKF JSON document from disk."""
    raw = Path(path).read_text(encoding="utf-8")
    doc = json.loads(raw)
    if not isinstance(doc, dict):
        raise ValueError("OKF root must be a JSON object")
    return doc


def validate(doc: Mapping[str, Any], *, stem: str | None = None) -> None:
    """Enforce OKF v1.0 structural rules used by the cross-language parity harness."""
    if doc.get("okf") != OKF_DIALECT:
        raise ValueError(f"expected okf '{OKF_DIALECT}', got {doc.get('okf')!r}")

    source_id = doc.get("source_id")
    if not isinstance(source_id, str) or not source_id:
        raise ValueError("missing non-empty source_id")
    if stem is not None and source_id != stem:
        raise ValueError(f"source_id {source_id!r} must equal filename stem {stem!r}")

    provenance = doc.get("provenance")
    if not isinstance(provenance, Mapping):
        raise ValueError("missing document provenance object")
    if provenance.get("source_id") != source_id:
        raise ValueError("provenance.source_id must equal top-level source_id")

    entities = doc.get("entities")
    if not isinstance(entities, Sequence) or isinstance(entities, (str, bytes)):
        raise ValueError("entities must be a JSON array")
    if len(entities) < 1:
        raise ValueError("entities must be non-empty")

    ids: set[str] = set()
    types: set[str] = set()
    for entity in entities:
        if not isinstance(entity, Mapping):
            raise ValueError("each entity must be an object")
        eid = entity.get("id")
        etype = entity.get("type")
        label = entity.get("label")
        if not isinstance(eid, str) or not eid:
            raise ValueError("entity missing id")
        if not isinstance(etype, str) or not etype:
            raise ValueError(f"entity {eid!r} missing type")
        if not isinstance(label, str) or not label.strip():
            raise ValueError(f"entity {eid!r} missing label")
        if eid in ids:
            raise ValueError(f"duplicate entity id {eid!r}")
        ids.add(eid)
        types.add(etype)

    missing = REQUIRED_SHARED_ENTITY_TYPES - types
    if missing:
        raise ValueError(f"missing required shared entity types: {sorted(missing)}")

    relations = doc.get("relations")
    if not isinstance(relations, Sequence) or isinstance(relations, (str, bytes)):
        raise ValueError("relations must be a JSON array")
    if len(relations) < 1:
        raise ValueError("relations must be non-empty")

    for rel in relations:
        if not isinstance(rel, Mapping):
            raise ValueError("each relation must be an object")
        src = rel.get("source")
        tgt = rel.get("target")
        rtype = rel.get("type")
        rprov = rel.get("provenance")
        if src not in ids:
            raise ValueError(f"relation source {src!r} not in entities")
        if tgt not in ids:
            raise ValueError(f"relation target {tgt!r} not in entities")
        if rtype not in ALLOWED_RELATION_TYPES:
            raise ValueError(f"relation type {rtype!r} not in OKF v1.0 set")
        if not isinstance(rprov, Mapping) or rprov.get("source_id") != source_id:
            raise ValueError("relation provenance.source_id must equal top-level source_id")


def emit(doc: Mapping[str, Any]) -> str:
    """Serialize a validated OKF document to pretty JSON with a trailing newline."""
    return json.dumps(doc, indent=2, ensure_ascii=False) + "\n"


def _stem_of(path: Path) -> str:
    name = path.name
    suffix = ".okf.json"
    if name.endswith(suffix):
        return name[: -len(suffix)]
    return path.stem


def main(argv: Sequence[str]) -> int:
    if len(argv) != 3 or argv[1] not in {"validate", "emit"}:
        sys.stderr.write(
            "usage: okf_adapter.py validate|emit <path.okf.json>\n"
        )
        return 2

    command, path_s = argv[1], argv[2]
    path = Path(path_s)
    if not path.is_file():
        sys.stderr.write(f"fixture not found: {path}\n")
        return 1

    try:
        doc = load(path)
        validate(doc, stem=_stem_of(path))
        if command == "validate":
            sys.stdout.write(f"OKF validate ok: {path} (source_id={doc['source_id']})\n")
            return 0
        sys.stdout.write(emit(doc))
        return 0
    except (OSError, json.JSONDecodeError, ValueError) as exc:
        sys.stderr.write(f"OKF adapter error: {exc}\n")
        return 1


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))

File: adapters/typescript/okf_adapter.ts
/**
 * OKF language adapter stub — TypeScript reference (C08 L75).
 *
 * Implements the language-agnostic contract in adapters/README.md:
 *   load(path) -> Record<string, unknown>
 *   validate(doc, stem?) -> void (throws on failure)
 *   emit(doc) -> string
 *
 * CLI:
 *   node --experimental-strip-types okf_adapter.ts validate <path.okf.json>
 *   node --experimental-strip-types okf_adapter.ts emit <path.okf.json>
 *
 * Hermetic: Node stdlib only. Not a Harbor / agent-eval harness.
 */

import { readFileSync } from "node:fs";
import { basename } from "node:path";

const OKF_DIALECT = "1.0";

const REQUIRED_SHARED_ENTITY_TYPES = new Set([
  "intent",
  "acceptance",
  "constraint",
  "resource",
  "state",
  "gate",
]);

const ALLOWED_RELATION_TYPES = new Set([
  "verified_by",
  "bounded_by",
  "grounds",
  "requires",
  "asserts",
]);

type JsonObject = Record<string, unknown>;

function load(path: string): JsonObject {
  const raw = readFileSync(path, "utf8");
  const doc: unknown = JSON.parse(raw);
  if (typeof doc !== "object" || doc === null || Array.isArray(doc)) {
    throw new Error("OKF root must be a JSON object");
  }
  return doc as JsonObject;
}

function asString(value: unknown): string | null {
  return typeof value === "string" ? value : null;
}

function asObject(value: unknown): JsonObject | null {
  return typeof value === "object" && value !== null && !Array.isArray(value)
    ? (value as JsonObject)
    : null;
}

function asArray(value: unknown): unknown[] | null {
  return Array.isArray(value) ? value : null;
}

function validate(doc: JsonObject, stem?: string): void {
  if (doc.okf !== OKF_DIALECT) {
    throw new Error(`expected okf '${OKF_DIALECT}', got ${String(doc.okf)}`);
  }

  const sourceId = asString(doc.source_id);
  if (!sourceId) {
    throw new Error("missing non-empty source_id");
  }
  if (stem !== undefined && sourceId !== stem) {
    throw new Error(
      `source_id ${JSON.stringify(sourceId)} must equal filename stem ${JSON.stringify(stem)}`,
    );
  }

  const provenance = asObject(doc.provenance);
  if (!provenance) {
    throw new Error("missing document provenance object");
  }
  if (provenance.source_id !== sourceId) {
    throw new Error("provenance.source_id must equal top-level source_id");
  }

  const entities = asArray(doc.entities);
  if (!entities) {
    throw new Error("entities must be a JSON array");
  }
  if (entities.length < 1) {
    throw new Error("entities must be non-empty");
  }

  const ids = new Set<string>();
  const types = new Set<string>();
  for (const item of entities) {
    const entity = asObject(item);
    if (!entity) {
      throw new Error("each entity must be an object");
    }
    const eid = asString(entity.id);
    const etype = asString(entity.type);
    const label = asString(entity.label);
    if (!eid) {
      throw new Error("entity missing id");
    }
    if (!etype) {
      throw new Error(`entity ${JSON.stringify(eid)} missing type`);
    }
    if (!label || !label.trim()) {
      throw new Error(`entity ${JSON.stringify(eid)} missing label`);
    }
    if (ids.has(eid)) {
      throw new Error(`duplicate entity id ${JSON.stringify(eid)}`);
    }
    ids.add(eid);
    types.add(etype);
  }

  const missing = [...REQUIRED_SHARED_ENTITY_TYPES].filter((t) => !types.has(t));
  if (missing.length > 0) {
    throw new Error(
      `missing required shared entity types: ${missing.sort().join(", ")}`,
    );
  }

  const relations = asArray(doc.relations);
  if (!relations) {
    throw new Error("relations must be a JSON array");
  }
  if (relations.length < 1) {
    throw new Error("relations must be non-empty");
  }

  for (const item of relations) {
    const rel = asObject(item);
    if (!rel) {
      throw new Error("each relation must be an object");
    }
    const src = rel.source;
    const tgt = rel.target;
    const rtype = rel.type;
    const rprov = asObject(rel.provenance);
    if (typeof src !== "string" || !ids.has(src)) {
      throw new Error(`relation source ${JSON.stringify(src)} not in entities`);
    }
    if (typeof tgt !== "string" || !ids.has(tgt)) {
      throw new Error(`relation target ${JSON.stringify(tgt)} not in entities`);
    }
    if (typeof rtype !== "string" || !ALLOWED_RELATION_TYPES.has(rtype)) {
      throw new Error(`relation type ${JSON.stringify(rtype)} not in OKF v1.0 set`);
    }
    if (!rprov || rprov.source_id !== sourceId) {
      throw new Error("relation provenance.source_id must equal top-level source_id");
    }
  }
}

function emit(doc: JsonObject): string {
  return `${JSON.stringify(doc, null, 2)}\n`;
}

function stemOf(path: string): string {
  const name = basename(path);
  const suffix = ".okf.json";
  if (name.endsWith(suffix)) {
    return name.slice(0, -suffix.length);
  }
  const dot = name.lastIndexOf(".");
  return dot >= 0 ? name.slice(0, dot) : name;
}

function main(argv: string[]): number {
  if (argv.length !== 4 || (argv[2] !== "validate" && argv[2] !== "emit")) {
    process.stderr.write(
      "usage: okf_adapter.ts validate|emit <path.okf.json>\n",
    );
    return 2;
  }

  const command = argv[2];
  const path = argv[3];

  try {
    const doc = load(path);
    validate(doc, stemOf(path));
    if (command === "validate") {
      process.stdout.write(
        `OKF validate ok: ${path} (source_id=${String(doc.source_id)})\n`,
      );
      return 0;
    }
    process.stdout.write(emit(doc));
    return 0;
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    process.stderr.write(`OKF adapter error: ${message}\n`);
    return 1;
  }
}

process.exit(main(process.argv));
